Task 4/Task 4/Task 4/FolderStateBuilder.cs

```diff
@@ -0,0 +1,170 @@
+                    catch
+                    {
+                        CreateSubFolders(file.FullPath);
+                        FileChanging(file);
+                    }
+                    break;
+                case FileWatcher.FileActions.Rename:
+                    FileRenaiming(file);
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// This method creates files with the help of information from example of
+        /// FileEventsInfo class.
+        /// </summary>
+        private static void FileCreation (FileEventsInfo file)
+        {
+            using (var stream = File.Create(file.FullPath)) { };
+            if (file.Content.Length != 0)
+            {
+                using (StreamWriter writer = new StreamWriter(file.FullPath, true, System.Text.Encoding.UTF8))
+                {
+                    writer.WriteLine(file.Content);
+                }
+            }
+            File.SetCreationTime(file.FullPath, file.LastChangesTime);
+            File.SetLastWriteTime(file.FullPath, file.LastChangesTime);
+        }
+
+        /// <summary>
+        /// This method creates subfolders, if they were created during observation mode.
+        /// It is necessary, because creating empty subfolder is not tracked, and doesn't 
+        /// record in log.
+        /// </summary>
+        private static void CreateSubFolders(string path)
+        {
+            string[] separateString = path.Split('\\');
+            string subFolderPath = String.Empty;
+            for (int i = 0; i < separateString.Length - 1; i++)
+            {
+                subFolderPath += separateString[i];
+                Directory.CreateDirectory(subFolderPath);
+                subFolderPath += @"\";
+            }
+        }
+
+        /// <summary>
+        /// This method removes text from file and record to file necessary stand text.
+        /// </summary>
+        private static void FileChanging(FileEventsInfo file)
+        {
+            File.WriteAllText(file.FullPath, String.Empty);
+            using (StreamWriter writer = new StreamWriter(file.FullPath, true, System.Text.Encoding.UTF8))
+            {
+                writer.Write(file.Content);
+            }
+            File.SetLastWriteTime(file.FullPath, file.LastChangesTime);
+        }
+
+        /// <summary>
+        /// This method contains logic of renaiming file.
+        /// </summary>
+        /// <param name="file"></param>
+        private static void FileRenaiming(FileEventsInfo file)
+        {
+            File.Delete(file.OldFullPath);
+            using (var stream = File.Create(file.FullPath)) { };
+            using (StreamWriter writer = new StreamWriter(file.FullPath, true, System.Text.Encoding.UTF8))
+            {
+                writer.WriteLine(file.Content);
+            }
+            File.SetLastWriteTime(file.FullPath, file.LastChangesTime);
+        }
```

Comment:
If it's only rename, you can use File.Move to "move" file to a new name. Then set last write time